prometheus/histogram.go

```diff
@@ -1835,3 +1835,143 @@ func (n *nativeExemplars) addExemplar(e *dto.Exemplar) {
 		n.exemplars = append(n.exemplars[:nIdx], append([]*dto.Exemplar{e}, append(n.exemplars[nIdx:rIdx], n.exemplars[rIdx+1:]...)...)...)
 	}
 }
+
+type constNativeHistogram struct {
+	desc *Desc
+	dto.Histogram
+	labelPairs []*dto.LabelPair
+}
+
+func validateCount(count uint64, negativeBuckets, positiveBuckets map[int]int64, zeroBucket uint64) error {
+	var bucktCountSum int64
+	for _, v := range positiveBuckets {
+		bucktCountSum += v
+	}
+	for _, v := range negativeBuckets {
+		bucktCountSum += v
+	}
+	bucktCountSum += int64(zeroBucket)
+	if bucktCountSum != int64(count) {
+		return fmt.Errorf("bucket sums do not add up to total count")
+	}
+	return nil
+}
+
+func NewConstNativeHistogram(
+	desc *Desc,
+	count uint64,
+	sum float64,
+	positiveBuckets,
+	negativeBuckets map[int]int64,
+	zeroBucket uint64,
+	nativeHistogramSchema int32,
+	nativeHistogramZeroThreshold float64,
+	createdTimestamp time.Time,
+	nativeExemplars []*dto.Exemplar,
+	labelValues ...string,
+) (Metric, error) {
+	if desc.err != nil {
+		return nil, desc.err
+	}
+	if err := validateLabelValues(labelValues, len(desc.variableLabels.names)); err != nil {
+		return nil, err
+	}
+	if nativeHistogramSchema > 8 || nativeHistogramSchema < -4 {
+		return nil, fmt.Errorf("invalid native histogram schema ")
+	}
+	if err := validateCount(count, negativeBuckets, positiveBuckets, zeroBucket); err != nil {
+		return nil, err
+	}
+
+	NegativeSpan, NegativeDelta := makeBucketsAny(negativeBuckets)
+	PositiveSpan, PositiveDelta := makeBucketsAny(positiveBuckets)
+	ret := &constNativeHistogram{
+		desc: desc,
+		Histogram: dto.Histogram{
+			CreatedTimestamp: timestamppb.New(createdTimestamp),
+			Schema:           &nativeHistogramSchema,
+			ZeroThreshold:    &nativeHistogramZeroThreshold,
+			Exemplars:        nativeExemplars,
+			SampleCount:      &count,
+			SampleSum:        &sum,
+
+			NegativeSpan:  NegativeSpan,
+			NegativeDelta: NegativeDelta,
+
+			PositiveSpan:  PositiveSpan,
+			PositiveDelta: PositiveDelta,
+
+			ZeroCount: proto.Uint64(zeroBucket),
+		},
+		labelPairs: MakeLabelPairs(desc, labelValues),
+	}
+	if *ret.ZeroThreshold == 0 && *ret.ZeroCount == 0 && len(ret.PositiveSpan) == 0 && len(ret.NegativeSpan) == 0 {
+		ret.PositiveSpan = []*dto.BucketSpan{{
+			Offset: proto.Int32(0),
+			Length: proto.Uint32(0),
+		}}
+	}
+	return ret, nil
+}
+
+func (h *constNativeHistogram) Desc() *Desc {
+	return h.desc
+}
+
+func (h *constNativeHistogram) Write(out *dto.Metric) error {
+	out.Histogram = &h.Histogram
+	out.Label = h.labelPairs
+	return nil
+}
+
+func makeBucketsAny(buckets map[int]int64) ([]*dto.BucketSpan, []int64) {
```

Comment:
hey, could you clarify what `Any` means here?

Reply:
Probably "positive or negative".
I would just call it `makeBuckets`.

Reply:
+1 for makeBuckets

Reply:
Hmm, makebuckets alreadyexists in the namespace. I could do `makeBucketfrommap`